File: 0238-product-of-array-except-self/0238-product-of-array-except-self.cpp
// space : o(n)
// time : o(n)
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        
        vector<int> res(nums.size() , -1);
        
        int prefix = 1;
        for(int i=0;i<nums.size();i++){
            res[i] = prefix;
            prefix = prefix* nums[i];
        }
        
        int postfix = 1;
        for(int i = nums.size()-1; i>=0 ; --i){
            res[i] = res[i]*postfix;
            postfix = postfix* nums[i];
       }
    return res;
        
    }    
};

/*
int main() {
    std::vector<int> nums = {1, 2, 3, 4};
    Solution solution;
    std::vector<int> result = solution.productExceptSelf(nums);
    
    for (int num : result) {
        std::cout << num << " ";
    }
    // Output: 24 12 8 6
    
    return 0;
}
*/

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        ## create an array of len that of input array(nums) and initialize every element as 1
        res = [1] * (len(nums))
        
        ## prefix case 
        prefix = 1
        for i in range(len(nums)):
            res[i] = prefix
            prefix = prefix*nums[i]
            
        ## postfix
        postfix = 1
        for i in  range(len(nums)-1 , -1, -1):
            res[i] = res[i]*postfix
            postfix = postfix*nums[i]
            
        return res    
        
        
# Example usage
#nums = [1, 2, 3, 4]
#solution = Solution()
#result = solution.productExceptSelf(nums)
#print(result)  # Output: [24, 12, 8, 6]#        